### ISSUE TITLE
CPU Speed 'Fastest' (and JIT combinations) don't work properly

### ISSUE THREAD

**Issue by blinkydoos** [midwan/amiberry#107]
When setting the CPU Speed to 'Fastest' and running the program SysInfo and doing a speed check it's detected as 7MHz. (JIT OFF).

With JIT ON, 7, 14 and 25MHz are all clocked at 7MHz (approximately 1279 Dhrystones).

**Comment by CypherXG** [comment]
My Test without JIT and fastest etc... Running on my Pi3, not overclocked
![dsc_0045](https://cloud.githubusercontent.com/assets/7895130/26284498/29b0fc38-3e3d-11e7-8c20-678d3c4c69e1.JPG)

![dsc_0046](https://cloud.githubusercontent.com/assets/7895130/26284502/2f1a203c-3e3d-11e7-8644-fe67d4fc09f0.JPG)

My Test with JIT and fastest etc... Running on my Pi3, not overclocked
![dsc_0047](https://cloud.githubusercontent.com/assets/7895130/26284507/4dc5b97e-3e3d-11e7-8860-134534e7d6b2.JPG)

![dsc_0048](https://cloud.githubusercontent.com/assets/7895130/26284508/51bfa0da-3e3d-11e7-92a9-97c949a145a0.JPG)

U see, i have no Problems with JIT etc.

Best regards... Markus

**Comment by blinkydoos** [comment]
Not sure why you have 68040 selected as it's not properly implemented yet. Your results say 68040 7MHz when you have JIT and Fastest selected, and why does your FPU say 'Not in Use' when you have it set at 'CPU internal' ? It seems to give different results on 68040. Could you try the test again with 68020 selected ?

@midwan Can you confirm any of this. I am using SDL2 version. I checked the Android version of UAE4Arm and it does the same thing so i'm not sure how CyperXG is having different results

**Comment by CypherXG** [comment]
Oh yeahh..there is an error. 
Is that what u are writing the Speed behind the CPU on the right side?

Everytime i do a "Hardreset" with uae4arm and boot from ADF Disk Sysinfo without any Workbench etc. 
Here are my results with all possible variants:
Have fun.... :-)

![grafik](https://cloud.githubusercontent.com/assets/7895130/26317392/9fb85eb4-3f17-11e7-9dfe-c644771d23d6.png)

Best regards...

**Comment by HoraceAndTheSpider** [comment]
Just from playing whdload games I am finding I am having to switch the cpu down to 7mhz on amiberry (it came from the desert, New Zealand story) when they should run fine on fast cpu's.

I am using 020 and "real" speed iirc. JIT seems to make no difference, although I often leave it on by default.

Also adding frame skip has no "throttling" to maintain a genuine speed.

**Comment by blinkydoos** [comment]
@CypherXG 
Thanks for the extensive list. As you can see there seems to be a lot more problems than I originally reported !!

My first guess would be that it's relying on 'Cycle-Accuracy' code which TomB never added to the emulator (and is sorely needed in a lot of games to stop them playing at turbo speed).

**Comment by midwan** [comment]
Yeap, the CPU code needs an update, to include cycle-exact modes and more (FPU code is mostly missing for example). This is a known issue and I was working on that slowly, but it's a big task and it takes time. :P

**Comment by blinkydoos** [comment]
Hi @midwan great to see you returning to this project. I've currently been using WinUAE but am always intrigued by the progress of this emulator for the Pi etc. I noticed that TomB has recently posted a couple of big updates to UAE4Arm. He's added features from WinUAE 3.3.0. I don't know if that includes Cycle Accuracy but he has improved the CPU Fastest settings etc. Maybe this could save you a ton of time by merging his updates to Amiberry ? Here's a link to his updates.

https://pyra-handheld.com/boards/threads/uae4arm.81153/

**Comment by midwan** [comment]
@blinkydoos 
I will definitely be looking into the changes TomB made, but perhaps it's better to wait a bit. He mentioned that this version is still in testing and not stable enough.

Also, the fact that he doesn't seem to have the source code on a standard repository, makes things difficult to merge (you'll have to detect changes in code manually, then compare manually with your code and pick what to merge in). Another reason to wait until his version is stable enough... :)

**Comment by midwan** [comment]
@blinkydoos 
All the latest changes (and more actually) have been integrated in the current sd1-dev and sdl2-dev branches. They will be merged to stable versions once we polish things up a bit (and add some more features).

Compatibility seems to be improved in several areas, speed as well. Some bugs still remain, but overall it's a step towards the right direction.

**Comment by TheMoleUK** [comment]
Has this improved in 2.25?

**Comment by midwan** [comment]
There have been bugfixes and improvements, especially in the JIT and JIT FPU area, but we haven't changed the emulated models.
Regarding sound settings, it depends on what you're emulating and what your personal preferences are. Try with the default settings and experiment from there :)

Finally, Fastest and Turbo modes change how much time Amiberry will give to CPU emulation. Turbo will give maximum time to CPU emulation (at the cost of reduced custom chipset emulation cycles), Fastest will give as much as possible while keeping the Custom Chips emulation going.

In short, only use Turbo for CPU-intensive work which requires no Custom Chipset emulation (e.g. Workbench applications, raytracing, etc).

**Comment by TheMoleUK** [comment]
Ah, thanks...
It’s all great fun...

**Comment by TheMoleUK** [comment]
Not exactly sure what JIT is...

**Comment by TheMoleUK** [comment]
The 030 could run at several speeds higher than 25MHz?

16, 20, 25, 33, 40, 50 MHz, except for MC68EC030 available in 25 and 40 MHz

**Comment by midwan** [comment]
Specific MHz make no sense under emulation, unless you're doing Cycle-Exact emulation (which we're not doing, because the current boards are now powerful enough to handle it).

**Comment by TheMoleUK** [comment]
I quite enjoy making an upgrade in emulation as best I can.  Given up on 040 and just played with 020 and 030 settings to create variants as best I can...
https://howlingpixel.com/i-en/Motorola_68030
